intros/TypeScript for the Haskell or ML Programmer.md

```diff
@@ -0,0 +1,658 @@
+    return t1.length > t2.length ? t1 : t2;
+}
+```
+
+TypeScript can usually infer type arguments from a call based on the
+type of the arguments, so type arguments are usually not needed.
+
+Because TypeScript is structural, it doesn't need type parameters as
+much as nominal systems. Specifically, they are not needed to make a
+function polymorphic. Type parameters should only be used to
+*propagate* type information, such as constraining parameters to be
+the same type:
+
+```ts
+function length<T extends ArrayLike<unknown>>(t: T): number {
+}
+
+function length(t: ArrayLike<unknown>): number {
+}
+```
+
+In the first `length`, T is not necessary; notice that it's only
+referenced once, so it's not being used to constrain the type of the
+return value or other parameters.
+
+TypeScript does not have higher kinded types, so the following is not legal:
+
+```ts
+function length<T extends ArrayLike<unknown>, U>(m: T<U>) {
+}
+```
+
+## Module system
+
+JavaScript's modern module syntax is a bit like Haskell's, except that
+any file with `import` or `export` is implicitly a module:
+
+```ts
+import { value, Type } from 'npm-package'
+import { other, Types } from './local-package'
+import * as prefix from '../lib/third-package'
+```
+
+You can also import commonjs modules &mdash; modules written using node.js'
+module system:
+
+```ts
+import f = require('single-function-package');
+```
+
+You can export with an export list:
+
+```ts
+export { f }
+
+function f() { return g() }
+function g() { } // g is not exported
+```
+
+Or by marking each export individually:
+
+```ts
+export function f { return g() }
+function g() { }
+```
+
+The latter style is more common but both are allowed, even in the same
+file.
+
+## `readonly` and `const`
+
+In JavaScript, mutability is the default, although it allows variable
+declarations with `const` to declare that the *reference* is
+immutable. The referent is still mutable:
+
+```js
+const a = [1,2,3];
+a.push(102); // ):
+a[0] = 101; // D:
+```
+
+TypeScript additionally has a `readonly` modifier for properties.
+
+```ts
+interface Rx {
+    readonly x: number;
+}
+let rx: Rx = { x: 1 };
+rx.x = 12; // error
+```
+
+It also ships with a mapped type `Readonly<T>` that makes
+all properties `readonly`:
+
+
+```ts
+interface X {
+    x: number;
+}
+let rx: Readonly<X> = { x: 1 };
+rx.x = 12; // error
+```
+
+And it has a specific `ReadonlyArray<T>` type that removes
+side-affecting methods and prevents writing to indices of the array,
+as well as special syntax for this type:
+
+```ts
+let a: ReadonlyArray<number> = [1, 2, 3];
+let b: readonly number[] = [1, 2, 3];
+a.push(102); // error
+b[0] = 101; // error
+```
+
+You can also use a const-assertion, which operates on arrays and
+object literals:
+
+```ts
+let a = [1, 2, 3] as const;
+a.push(102); // error
+a[0] = 101; // error
+```
+
+However, none of these options are the default, so they are not
+consistently used in TypeScript code.
+
+# Further reading
+
```

Comment:
Currying 🥘

Reply:
Done, as "point-free programming", so as to warn people away from pipe, partial, compose, and more complex relatives.